### ISSUE TITLE
Unknown type from struct.unpack

### ISSUE THREAD

**Issue by mikeshardmind** [microsoft/pyright#4727]
**Describe the bug**
pyright does not understand struct.unpack/struct.unpack_from

It isn't feasible to handle this solely off of what's supported in the current typing ecosystem, but the [format strings in struct are rigidly defined ](https://docs.python.org/3/library/struct.html#format-characters)in such a way that literal format strings should allow for type inference.
As such, I'm hoping that tooling can bridge the gap, at least in the case of literal format strings.

**To Reproduce**
Use `struct.unpack` / `struct.unpack_from`

**Expected behavior**
Type checking to understand the type returned by `struct.unpack` / `struct.unpack_from`

**Screenshots or Code**
```python
def _get_data_version(b: bytes, /) -> int:
    return struct.unpack("!b", b)[0]
```

In this particular case, `struct.unpack` may only return a single element tuple with that element being an integer that fits in a single byte or raise an exception. While pyright allows this, it doesn't correctly understand the type, only understanding it to be a tuple.

The ideal scenario would also handle this case:
```py
def _extract_length_prefixed_array(b: bytes, offset: int) -> tuple[int, ...]:
    (length, ) = struct.unpack_from("!b", b, offset)
    return struct.unpack_from("!b%dQ" % length, b, offset + 1)
```

As it is possible to tell from the format string that the only thing being formatted is a quantifier and that all elements in the resulting tuple must be an integer, but this might be a stretch for what can reasonably be supported.

Note that while pyright allows what is above, it allows it in a way that removes any benefit to having this code be type-checked.
If this were a more-involved function (this is only a minimal reproduction), this could have subtle type errors here. (shown by taking the solution used below and adding reveal_type use)
![image](https://user-images.githubusercontent.com/22035876/222890699-58af249b-9d9a-49f3-9021-187e89686026.png)

**VS Code extension or command-line**
Indirectly through pylance as a language server with strict type checking enabled, pylance version v2023.3.10

**Additional context**

Currently, this is being solved in code as 
```py
def _get_data_version(b: bytes, /) -> int:
    (r,) = struct.unpack("!b", b)
    assert isinstance(r, int)
    return r
```

to have minimal impact while retaining the benefits of strict type checking, but the minimal impact relies on users running python with optimizations enabled, which is unfortunately (from various analytics collected previously) not a default option for most users. I'm currently mulling over options here as there are various places where struct handling should be as low-overhead as reasonably possible. The best case would involve direct support in pyright, with the second best case possibly ending up as a type ignore (unfortunately)

**Comment by erictraut** [comment]
Pyright is working as designed here.

The typeshed stubs define `struct.unpack` as follows:

```python
def unpack(__format: str | bytes, __buffer: ReadableBuffer) -> tuple[Any, ...]: ...
```

Pyright is enforcing the type information it's provided. It has no knowledge of the internals of this particular method (or, generally, any other method or function). If you think that the typeshed type definition is in error here, please file a bug with the typeshed project.

If the typeshed type definition is correct, then you'll need to use some technique such as the one in your workaround above.

**Comment by mikeshardmind** [comment]
Hi, thanks for the confirmation on this. 

It's unfortunate because the typeshed definition is as correct as possible given what is supported by typing. Yet, there is more detailed concrete type information available for static analysis here from the struct format string. When the string is Literal or composed of Literals in a predictable way, as well as when a string meeting those criteria is formatted only with numbers, the type information can be known from just that.

I guess I should read this response as such a feature being out of scope for pyright, but I'm a little disappointed by this.

I'm also unsure of even how this could be incorporated in typing directly rather than as part of any analysis tool's understanding of the standard lib's behavior without some rather significant changes. struct format strings can be literal, composed of literals, or dynamic. even in the case of them being literals, no typing support exists for generating type information from these as it wouldn't be considered static anymore.

It would be a little less painful if there was a good (read: zero additional code/runtime cost) way to coerce types for cases that don't play nicely with python's type system (such as a comment like `# type: tuple[int, ...]` being taken as the type)

assertions work for individual elements but work poorly in many others.

**Comment by alexchandel** [comment]
It would be nice for Pyright to special-case `struct.unpack`, because it's in the Python standard library and because it has an exact format.  Typeshed stubs can't easily capture the dependent type of the output, but Pyright could algorithmically work out literal formats, like that `struct.unpack('UUf', data)` returns a `tuple[int, int, float]`.

**Comment by alexchandel** [comment]
I think it's time to revisit this decision, since astral's `ty` infers return types for `struct.unpack` (see astral-sh/ruff#22562 and astral-sh/ruff#23130).

**Comment by mikeshardmind** [comment]
@alexchandel https://github.com/microsoft/pyright/pull/11494 was merged a week ago. It still doesn't handle some cases that could be handled (it only handles literals, where it is possible to handle some formatted versions statically), and in those cases, results in a gradual type that can't be coerced for free under current type system rules, the latter continuing to be a problem with the typeshed + type system rules, not with pyright.

For users who ran into this before, this is probably a significant improvement, but I haven't tried using it.
